aten/src/ATen/core/jit_type.h

```diff
@@ -541,10 +543,11 @@ struct CAFFE2_API FutureType : public SingleElementType<TypeKind::FutureType, Fu
 
 struct TupleType;
 using TupleTypePtr = std::shared_ptr<TupleType>;
+using OptNameList = c10::optional<std::vector<std::string>>;
 // This type represents a Tuple
 struct CAFFE2_API TupleType : public Type {
-  static TupleTypePtr create(std::vector<TypePtr> types) {
-    return TupleTypePtr(new TupleType( std::move(types) )); // NOLINT(modernize-make-shared)
+  static TupleTypePtr create(std::vector<TypePtr> types, OptNameList names=c10::nullopt) {
```

Comment:
Isn't the rule, named tuple is a subtype of unnamed tuple, but not vice versa? Because a named tuple can be used in any context where an unnamed tuple is expected.

OTOH, this is just if I were purely designing for the Python frontend type system. Aren't these types for our internal core representation? I would have naively expected the distinction to get erased by that point, but I haven't read this patch.

Reply:
Yeah, I think that is right. In which case, == needs to be adjusted to return false if the names do not match, and isSubtypeOf needs to be adjust so that `N <: U`. It is not erased because it needs to get propagated around with the type tuple.